### PR TITLE
Issue #1870 - Fix extraneous '&' in compartment search response links

### PR DESCRIPTION
**Pull request by tbieste**
Fixes extraneous '&' characters in compartment search response links.

Signed-off-by: Troy Biesterfeld <tbieste@us.ibm.com>

### CODE REVIEW

**Review by michaelwschroeder — APPROVED** [review]

**Comment by michaelwschroeder** [comment]
LGTM

File: fhir-search/src/main/java/com/ibm/fhir/search/uri/UriBuilder.java [hunk 1]
```diff
@@ -107,6 +99,21 @@ public String toSearchSelfUri() throws URISyntaxException {
...
     }
 
+    /**
+     * Appends the search parameters to the query string, if joining them together results in a non-empty string.
+     */
+    private void appendSearchParameters() {
+        if (!context.getSearchParameters().isEmpty()) {
+            Predicate<String> stringIsEmpty = String::isEmpty;
+            String searchParameters = context.getSearchParameters().stream().map(p -> serializeSearchParmToQueryString(p))
+                    .filter(stringIsEmpty.negate()).collect(Collectors.joining(SearchConstants.AND_CHAR_STR));
```
Comment (lmsurpre) [inline fhir-search/src/main/java/com/ibm/fhir/search/uri/UriBuilder.java:109]: minor style suggestion
```suggestion
            String searchParameters = context.getSearchParameters().stream()
                    .map(p -> serializeSearchParmToQueryString(p))
                    .filter(stringIsEmpty.negate())
                    .collect(Collectors.joining(SearchConstants.AND_CHAR_STR));
```

**Review by lmsurpre — APPROVED** [review]
LGTM

File: fhir-search/src/main/java/com/ibm/fhir/search/uri/UriBuilder.java (hunk 1 above)
Comment (tbieste) [reply fhir-search/src/main/java/com/ibm/fhir/search/uri/UriBuilder.java]: Updated.

**Comment by tbieste** [comment]
Comments addressed.  Ready to merge.

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/SearchTest.java [hunk 2]
```diff
@@ -639,6 +639,9 @@ public void testSearchObservationWithPatientCompartment() {
         Bundle bundle = response.readEntity(Bundle.class);
         assertNotNull(bundle);
         assertTrue(bundle.getEntry().size() >= 1);
+        // verify link does not include consecutive '&' characters
+        assertTrue(bundle.getLink().size() >= 1);
+        assertTrue(!bundle.getLink().get(0).getUrl().getValue().contains("&&"));
```
Comment (prb112) [inline fhir-server-test/src/test/java/com/ibm/fhir/server/test/SearchTest.java:644]: For Your Awareness... assertFalse is a good choice here.

Comment (tbieste) [reply fhir-server-test/src/test/java/com/ibm/fhir/server/test/SearchTest.java]: Updated.

**Pull request merged**